chain/sub/incoming.go

```diff
@@ -445,23 +449,202 @@ func recordFailure(ctx context.Context, metric *stats.Int64Measure, failureType
 	stats.Record(ctx, metric.M(1))
 }
 
+type peerMsgInfo struct {
+	peerID    peer.ID
+	lastCid   cid.Cid
+	rateLimit *ratelimit.Window
+	mutex     sync.Mutex
+}
+
 type IndexerMessageValidator struct {
 	self peer.ID
+
+	peerCache *lru.TwoQueueCache
+	chainApi  full.ChainModuleAPI
+	stateApi  full.StateModuleAPI
 }
 
-func NewIndexerMessageValidator(self peer.ID) *IndexerMessageValidator {
-	return &IndexerMessageValidator{self: self}
+func NewIndexerMessageValidator(self peer.ID, chainApi full.ChainModuleAPI, stateApi full.StateModuleAPI) *IndexerMessageValidator {
+	peerCache, _ := lru.New2Q(1024)
+
+	return &IndexerMessageValidator{
+		self:      self,
+		peerCache: peerCache,
+		chainApi:  chainApi,
+		stateApi:  stateApi,
+	}
 }
 
 func (v *IndexerMessageValidator) Validate(ctx context.Context, pid peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
 	// This chain-node should not be publishing its own messages.  These are
-	// relayed from miner-nodes or index publishers.  If a node appears to be
-	// local, reject it.
+	// relayed from market-nodes.  If a node appears to be local, reject it.
 	if pid == v.self {
 		log.Warnf("refusing to relay indexer message from self")
 		stats.Record(ctx, metrics.IndexerMessageValidationFailure.M(1))
 		return pubsub.ValidationReject
 	}
+	originPeer := msg.GetFrom()
+	if originPeer == v.self {
```

Comment:
this will render your node incapable of publishing, the validator runs for all messages, including yours.

You should check the ReceivedFrom field and if it is locally published, let it through.

Reply:
it's not the lotus daemon that produces these messages, but rather the markets libp2p host.
The markets process in MRA shouldn't have access to the daemon identity key.

The validation structure here is that markets signs it, and lotus daemons relaying the gossip message will look at the lotus miner id that's in the message, look on chain to see the markets libp2p host delegated by that miner, and confirm that the message is signed by that markets libp2p host.

Reply:
thats fine, all i am saying is that you should look at yhe ReceivedFrom and not drop lically published messages.

Isnt the same code used by the miner? How does it enter the overlay?

Reply:
it will get sent out on the markets libp2p host which is passed to the index-provider library.

That library will default to only allowing locally generated messages and not relaying any gossip messages.

We're keeping the markets libp2p host as a protected peer to the lotus daemon host, and were worried about cases where a malicious peer was able to connect to the markets process and then send gossip messages that got relayed out from the lotus daemon and affected it's score in the mesh. It seems safer to say the markets libp2p host shouldn't become part of the larger lotus mesh and will only originate it's local gossip messages.

Reply:
so you forward the message to the miner which then publishes to the overlay? Basically some form of delagated publishing.

Reply:
basically i am a little confused as to who publishes (and signs/gets the origin id) here.

Reply:
basically i want to avoid the footgun of accidentally dropping your own messages.

Reply:
The markets libp2p host publishes. its only expected connected peer is it's associated lotus daemon which will relay it into the rest of the relay.

The signer/origin will be of the markets libp2p host, not of the lotus daemon itself.

Reply:
does the market process have the miner peer id?

ok, fair enough.

Reply:
Since this validation only happens for chain nodes (`modules.RelayIndexerMessages` is only a `ChainNode` option) that are relaying a message, and a node that published an indexer message should never relay such a message, this check seems OK.

However, it is really necessary or better to just not worry about checking this?

Reply:
well, lets keep it for now. But i would suggest making it an ignore instead of reject.

Reply:
Ignored now